src/docs/runtime-metadata.md

```diff
@@ -0,0 +1,255 @@
+2 - using an accessor
+
+Within accessors, there are two classes of accessor:
+1 - compiler-generated accessors (for all nominal types)
+2 - runtime-provided accessors (for non-nominal types - tuples, closures, etc.)
+
+Generally speaking, it's probably best for us to use the accessors and if we choose, to cache those values as static fields in bound types.
+
+The signature of a type metadata accessor is a request followed by n TypeMetadata objects, one for each generic type:
+```csharp
+public static TypeMetadata Accessor(TypeMetadataRequest request [, TypeMetadata specialization0, TypeMetadata specialization1, ...])
+{
+
+}
+```
+The request is the state of the resulting type metadata object. This is documented [here](https://github.com/swiftlang/swift/blob/9fed7159e9419e27f081364f7501f6fc7592f8c2/include/swift/ABI/MetadataValues.h#L2392), but for our needs, I don't anticipate any use but `Complete` which is 0.
+
+## Implementation
+
+```csharp
+namespace Swift.Runtime;
+
+[Flags]
+public enum TypeMetadataFlags {
+    MetadataKindIsNonType = 0x400,
+    MetadataKindIsNonHeap = 0x200,
+    MetadataKindIsRuntimePRivate = 0x100,
+}
+
+[Flags]
+public enum TypeMetadataRequest {
+    Complete = 0,
+    NonTransitiveComplete = 1,
+    LayoutComplete = 0x3f,
+    Abstract = 0xff;
+    IsNotBlocking = 0x100,
+}
+
+public enum TypeMetadataKind {
+    None = 0,
+    Struct = 0 | MetadataFlags.MetadataKindIsNonHeap,
+    Enum = 1 | MetadataFlags.MetadataKindIsNonHeap,
+    Optional = 2 | MetadataFlags.MetadataKindIsNonHeap,
+    ForeignClass = 3 | MetadataFlags.MetadataKindIsNonHeap,
+    Opaque = 0 | MetadataFlags.MetadataKindIsRuntimePrivate | MetadataFlags.MetadataKindIsNonHeap,
+    Tuple = 1 | MetadataFlags.MetadataKindIsRuntimePrivate | MetadataFlags.MetadataKindIsNonHeap,
+    Function = 2 | MetadataFlags.MetadataKindIsRuntimePrivate | MetadataFlags.MetadataKindIsNonHeap,
+    Protocol = 3 | MetadataFlags.MetadataKindIsRuntimePrivate | MetadataFlags.MetadataKindIsNonHeap,
+    Metatype = 4 | MetadataFlags.MetadataKindIsRuntimePrivate | MetadataFlags.MetadataKindIsNonHeap,
+    ObjCClassWrapper = 5 | MetadataFlags.MetadataKindIsRuntimePrivate | MetadataFlags.MetadataKindIsNonHeap,
+    ExistentialMetatype = 6 | MetadataFlags.MetadataKindIsRuntimePrivate | MetadataFlags.MetadataKindIsNonHeap,
+    HeapLocalVariable = 0 | MetadataFlags.MetadataKindIsNonType,
+    HeapGenericLocalVariable = 0 | MetadataFlags.MetadataKindIsNonType | MetadataFlags.MetadataKindIsRuntimePrivate,
+    ErrorObject = 1 | MetadataFlags.MetadataKindIsNonType | MetadataFlags.MetadataKindIsRuntimePrivate,
+    Class = 0x800 // not really, but it's reasonable under the circumstances
+                // This number looks arbitrary, but the swift source says that they're never going to
+                // go over 0x7ff for predefined/enumerated metadata kind values
+}
+
+public record struct TypeMetadata
+{
+    const nint kMaxDiscriminator = 0x7ff;
+    NativeHandle handle;
```

Comment:
Why do we need `NativeHandle` instead of `IntPtr`?

Reply:
I mentioned that earlier in the docs. `IntPtr` is a synonym for `nint` and I want to avoid using it even though it technically could be used here.